arangod/Aql/WindowExecutor.cpp

```diff
@@ -0,0 +1,439 @@
+////////////////////////////////////////////////////////////////////////////////
+/// DISCLAIMER
+///
+/// Copyright 2020 ArangoDB GmbH, Cologne, Germany
+///
+/// Licensed under the Apache License, Version 2.0 (the "License");
+/// you may not use this file except in compliance with the License.
+/// You may obtain a copy of the License at
+///
+///     http://www.apache.org/licenses/LICENSE-2.0
+///
+/// Unless required by applicable law or agreed to in writing, software
+/// distributed under the License is distributed on an "AS IS" BASIS,
+/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+///
+/// @author Simon Grätzer
+////////////////////////////////////////////////////////////////////////////////
+
+#include "WindowExecutor.h"
+
+#include "Aql/Aggregator.h"
+#include "Aql/AqlCall.h"
+#include "Aql/AqlValue.h"
+#include "Aql/ExecutionNode.h"
+#include "Aql/InputAqlItemRow.h"
+#include "Aql/OutputAqlItemRow.h"
+#include "Aql/RegisterInfos.h"
+#include "Aql/RegisterPlan.h"
+#include "Aql/SingleRowFetcher.h"
+
+#include <utility>
+
+using namespace arangodb;
+using namespace arangodb::aql;
+
+namespace {
+static const AqlValue EmptyValue;
+}
+
+WindowExecutorInfos::WindowExecutorInfos(WindowBounds const& bounds, RegisterId rangeRegister,
+                                         std::vector<std::string>&& aggregateTypes,
+                                         std::vector<std::pair<RegisterId, RegisterId>>&& aggregateRegisters,
+                                         QueryWarnings& w, velocypack::Options const* opts)
+    : _bounds(bounds),
+      _rangeRegister(rangeRegister),
+      _aggregateTypes(aggregateTypes),
+      _aggregateRegisters(aggregateRegisters),
+      _warnings(w),
+      _vpackOptions(opts) {
+  TRI_ASSERT(!aggregateRegisters.empty());
+}
+
+WindowBounds const& WindowExecutorInfos::bounds() const { return _bounds; }
+
+RegisterId WindowExecutorInfos::rangeRegister() const { return _rangeRegister; }
+
+std::vector<std::pair<RegisterId, RegisterId>> WindowExecutorInfos::getAggregatedRegisters() const {
+  return _aggregateRegisters;
+}
+
+std::vector<std::string> WindowExecutorInfos::getAggregateTypes() const {
+  return _aggregateTypes;
+}
+
+QueryWarnings& WindowExecutorInfos::warnings() const { return _warnings; }
+
+velocypack::Options const* WindowExecutorInfos::getVPackOptions() const {
+  return _vpackOptions;
+}
+
+BaseWindowExecutor::AggregatorList BaseWindowExecutor::createAggregators(
+    WindowExecutor::Infos const& infos) {
+  AggregatorList aggregators;
+
+  TRI_ASSERT(!infos.getAggregateTypes().empty());
+  if (infos.getAggregateTypes().empty()) {
+    // no aggregate registers. this means we'll only count the number of items
+    THROW_ARANGO_EXCEPTION(TRI_ERROR_INTERNAL);
```

Comment:
If for whatever reasons this happens, we should add a more descriptive error message, e.g.
```suggestion
    THROW_ARANGO_EXCEPTION_MESSAGE(TRI_ERROR_INTERNAL, "no aggregators found in WindowExecutor");
```